File: day14/main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"math"
	"os"
	"strings"
)

func initPairCounts(template string) map[string]int {

	pairCounts := make(map[string]int)
	for i := 0; i < len(template)-1; i++ {
		pairCounts[template[i:i+2]] += 1
	}

	return pairCounts
}

func updatePairCounts(pairCounts map[string]int, rules map[string]string) map[string]int {

	newPairCounts := make(map[string]int)
	for k, v := range pairCounts {
		newPairCounts[k] = v
	}
	for pair, ins := range rules {
		newPairCounts[string(pair[0])+ins] += pairCounts[pair]
		newPairCounts[ins+string(pair[1])] += pairCounts[pair]
		newPairCounts[pair] -= pairCounts[pair]
	}
	return newPairCounts
}

// getElementCounts counts all elements in the polymer using pair counts
// It does this by counting the second item in each pair. Since the first element
// in the polymer is not the second item of a pair, it needs to be added once
func getElementCounts(polymerTemplate string, pairCounts map[string]int) map[string]int {

	counts := make(map[string]int)
	for k, v := range pairCounts {
		counts[string(k[1])] += v
	}

	counts[string(polymerTemplate[0])] += 1

	return counts
}

func getMinAndMaxDiff(counts map[string]int) int {

	max_v := 0
	min_v := math.MaxInt64
	for _, v := range counts {
		if v > max_v {
			max_v = v
		}
		if v < min_v {
			min_v = v
		}
	}
	return max_v - min_v
}

func run(steps int, polymerTemplate string, rules map[string]string) int {
	pairCounts := initPairCounts(polymerTemplate)

	for i := 0; i < steps; i++ {
		pairCounts = updatePairCounts(pairCounts, rules)
	}

	counts := getElementCounts(polymerTemplate, pairCounts)
	return getMinAndMaxDiff(counts)
}

func main() {
	var file string
	flag.StringVar(&file, "infile", "input", "Input file")
	flag.Parse()

	content, err := os.ReadFile(file)
	if err != nil {
		log.Fatal(err)
	}

	lines := strings.Split(strings.Trim(string(content), "\n"), "\n")

	polymerTemplate := lines[0]
	rules := make(map[string]string)
	for _, line := range lines[2:] {
		val := strings.Split(line, " -> ")
		rules[val[0]] = val[1]
	}
	fmt.Println(run(10, polymerTemplate, rules))
	fmt.Println(run(40, polymerTemplate, rules))
}

File: day14/main.py
from argparse import ArgumentParser
from collections import Counter
from pathlib import Path


def init_pair_counts(template):

    pair_counts = Counter()
    for i in range(len(polymer_template[:-1])):
        pair_counts[polymer_template[i:i+2]] += 1

    return pair_counts


def update_pair_counts(pair_counts, rules):
    """
    When an element is inserted between a pair, that pair is eliminated and the new
    pairs are added.
    """

    # Keep a separate copy of the counter because the rules are applied at the same time
    new_pair_counts = Counter(pair_counts)
    for pair, ins in rules:
        new_pair_counts[pair[0]+ins] += pair_counts[pair]
        new_pair_counts[ins+pair[1]] += pair_counts[pair]
        new_pair_counts[pair] -= pair_counts[pair]
    return new_pair_counts


def run(steps, polymer_template, rules):
    pair_counts = init_pair_counts(polymer_template)

    for _ in range(steps):
        pair_counts = update_pair_counts(pair_counts, rules)

    return get_element_counts(pair_counts, polymer_template[0])


def get_element_counts(pair_counts, first_char):
    """
    Count each second element in a pair. By doing this we miss the first element of the
    polymer, so add 1 for it.
    """

    counts = Counter()
    for k, v in pair_counts.items():
        counts[k[1]] += v

    counts[first_char] += 1

    return counts


if __name__ == "__main__":

    parser = ArgumentParser()
    parser.add_argument("-i", "--infile", required=False, type=Path, default=Path("input"))
    args = parser.parse_args()

    lines = Path(args.infile).read_text().rstrip("\n").split("\n")
    polymer_template = lines[0]
    rules = [rule.split(" -> ") for rule in lines[2:]]

    counts = run(10, polymer_template, rules)
    print(counts.most_common()[0][1]-counts.most_common()[-1][1])

    counts = run(40, polymer_template, rules)
    print(counts.most_common()[0][1]-counts.most_common()[-1][1])
